File: downloads/code/LeetCode/Python/Combination-Sum2.py
class Solution(object):                                                                          
    def comb_rec(self, c, target, ret, tmp,index):
        if target == 0 and tmp not in ret:
            ret.append(tmp[:])
            return
        if target < 0:
            return

        for i in range(index,len(c)):
            if target-c[i] < 0:
                return
            tmp.append(c[i])
            self.comb_rec(c,target-c[i],ret,tmp,i+1)
            tmp.pop()

    def combinationSum2(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        candidates = sorted(candidates)
        ret = []
        if not candidates:
            return ret
        self.comb_rec(candidates,target,ret,[],0)
        return ret

File: downloads/code/LeetCode/Java/Combination-Sum2.java
public class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        List<List<Integer>> ret = new ArrayList<List<Integer>> ();
        List<Integer> tmp = new ArrayList<Integer>();
        dfs(ret, tmp, candidates, target, 0);
        return ret;
    }

    public void dfs(List<List<Integer>> ret, List<Integer> tmp, int[] candidates, int target, int index){
        if(target==0){
            ret.add(new ArrayList<Integer>(tmp));
            return;
        }
        if(target<0) return;

        for(int i=index;i<candidates.length;i++){
            if(i>index && candidates[i]==candidates[i-1]) continue;
            tmp.add(candidates[i]);
            dfs(ret, tmp, candidates, target-candidates[i], i+1);
            tmp.remove(tmp.size()-1);
        
        }
    }
}
